In [ ]:
# importing major liberacy
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# importing warning remover
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# Sklearm major liberies
from sklearn.model_selection import train_test_split
from sklearn.ensemble import StackingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier

# Imported major liberies from sklearn for training and testing algorithm

In [ ]:
df = pd.read_csv('diabetes.csv')

In [ ]:
df

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
...,...,...,...,...,...,...,...,...,...
763,10,101,76,48,180,32.9,0.171,63,0
764,2,122,70,27,0,36.8,0.340,27,0
765,5,121,72,23,112,26.2,0.245,30,0
766,1,126,60,0,0,30.1,0.349,47,1


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
df.isnull().sum()

,0
Pregnancies,0
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [ ]:
df.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [ ]:
x = df.drop('Outcome', axis=1)
y = df['Outcome']

In [ ]:
le = LabelEncoder()
y_encoded = le.fit_transform(y)

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(x,y_encoded,test_size=0.2,random_state=42,stratify=y_encoded)

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(614, 8)
(154, 8)
(614,)
(154,)


In [ ]:
base_learners = [
    ('dt', DecisionTreeClassifier(random_state=42)),
    ('svc', SVC(probability=True, kernel='rbf', random_state=42)),
    ('lr', LogisticRegression(max_iter=1000))
]

In [ ]:
meta_learner = LogisticRegression(max_iter=1000)

In [ ]:
stacking_clf = StackingClassifier(
    estimators=base_learners,
    final_estimator=meta_learner,
    cv=5
)

In [ ]:
stacking_clf.fit(x_train, y_train)

StackingClassifier(cv=5,
                   estimators=[('dt', DecisionTreeClassifier(random_state=42)),
                               ('svc', SVC(probability=True, random_state=42)),
                               ('lr', LogisticRegression(max_iter=1000))],
                   final_estimator=LogisticRegression(max_iter=1000))

In [ ]:
y_pred = stacking_clf.predict(x_test)

In [ ]:
accuracy = accuracy_score(y_test,y_pred)

In [ ]:
accuracy

0.7337662337662337

| Algorithm                        | Accuracy   |                                                                                                                                                                                                                                                                                            |
| -------------------------------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Stacking Classifier**          | **73.38%** |

#Description
An ensemble learning technique that combines predictions from multiple base models (**Decision Tree, SVM, and Logistic Regression**) and uses a meta-model (**Logistic Regression**) to make the final prediction. It leverages the strengths of different algorithms to improve overall performance. |

In [ ]:
scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [ ]:
# ramdom forest classifier
rf_model = RandomForestClassifier(
    n_estimators=100,     # number of trees
    max_depth=None,       # let trees grow fully
    random_state=42,

)

In [ ]:
rf_model.fit(x_train,y_train)

RandomForestClassifier(random_state=42)

In [ ]:
y_pred = rf_model.predict(x_test)

In [ ]:
accuracy = accuracy_score(y_test,y_pred)

In [ ]:
accuracy

0.7597402597402597

| Algorithm                        | Accuracy   |                                                                                                                                                                                                                                                                                            |
| -------------------------------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Random Forest Classifier**     | **75.97%** |

# Description
An ensemble algorithm that creates multiple decision trees and combines their predictions using majority voting. It reduces overfitting and improves generalization compared to a single decision tree.

In [ ]:
# ada boost
ada_model = AdaBoostClassifier(n_estimators=100, random_state=42)

In [ ]:
ada_model.fit(x_train,y_train)

AdaBoostClassifier(n_estimators=100, random_state=42)

In [ ]:
y_pred = ada_model.predict(x_test)

In [ ]:
accuracy = accuracy_score(y_test,y_pred)

In [ ]:
accuracy

0.7922077922077922

| Algorithm                        | Accuracy   |                                                                                                                                                                                                                                                                                            |
| -------------------------------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **AdaBoost Classifier**     | **79.22%** |

---



# Description
A boosting algorithm that sequentially trains weak learners and gives more weight to previously misclassified instances. It focuses on difficult observations and improves prediction accuracy over iterations.

In [ ]:
# GradientBoost
gb_model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, random_state=42)


In [ ]:
gb_model.fit(x_train,y_train)

GradientBoostingClassifier(random_state=42)

In [ ]:
y_pred = gb_model.predict(x_test)

In [ ]:
accuracy = accuracy_score(y_test,y_pred)

In [ ]:
accuracy

0.7532467532467533

| Algorithm                        | Accuracy   |                                                                                                                                                                                                                                                                                            |
| -------------------------------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Gradient Boosting Classifier**     | **75.32%** |

---



# Description
A boosting technique that builds trees sequentially, where each new tree corrects the errors made by previous trees. It optimizes performance by minimizing the loss function through gradient descent.

In [ ]:
# XG boost
xgb_model = XGBClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, use_label_encoder=False, eval_metric='mlogloss', random_state=42)

In [ ]:
xgb_model.fit(x_train,y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=3, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=100, n_jobs=None,
              num_parallel_tree=None, ...)

In [ ]:
y_pred = xgb_model.predict(x_test)

In [ ]:
accuracy = accuracy_score(y_test,y_pred)

In [ ]:
accuracy

0.7532467532467533

| Algorithm                        | Accuracy   |                                                                                                                                                                                                                                                                                            |
| -------------------------------- | ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **XGBoost Classifier**     | **75.32%** |

---



# Description
An advanced implementation of gradient boosting that includes regularization, parallel processing, and optimized tree building. It is known for high performance and efficiency on structured datasets.

# Model Performance Ranking

**AdaBoost Classifier** – 79.22% (Best Performing Model)

**Random Forest Classifier** – 75.97%

**Gradient Boosting Classifier** – 75.32%

**XGBoost Classifier** – 75.32%

**Stacking Classifier** – 73.38%

# Business Conclusion
AdaBoost achieved the highest accuracy (79.22%) and is the recommended model for deployment.
Random Forest provides a good balance between accuracy and interpretability.
Gradient Boosting and XGBoost delivered similar performance, indicating that further hyperparameter tuning may be required to unlock XGBoost's full potential.
Stacking Classifier did not outperform the individual ensemble models, suggesting that the selected base learners may not be sufficiently diverse for effective stacking.